ddl/partition.go

```diff
@@ -1940,27 +1953,139 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
 		return ver, errors.Trace(dbterror.ErrPartitionMgmtOnNonpartitioned)
 	}
 
-	newPartitions := make([]model.PartitionDefinition, 0, len(oldIDs))
-	for _, oldID := range oldIDs {
+	switch job.SchemaState {
+	case model.StatePublic:
+		// Step1: generate new partition ids
+		newDefinitions := make([]model.PartitionDefinition, 0, len(pi.Definitions))
 		for i := 0; i < len(pi.Definitions); i++ {
-			def := &pi.Definitions[i]
-			if def.ID == oldID {
-				pid, err1 := t.GenGlobalID()
-				if err1 != nil {
-					return ver, errors.Trace(err1)
+			newDef := pi.Definitions[i].Clone()
+			for _, oldID := range oldIDs {
+				if newDef.ID == oldID {
+					pid, err1 := t.GenGlobalID()
+					if err1 != nil {
+						return ver, errors.Trace(err1)
+					}
+					newDef.ID = pid
+					break
 				}
-				def.ID = pid
-				// Shallow copy only use the def.ID in event handle.
-				newPartitions = append(newPartitions, *def)
-				break
 			}
+			newDefinitions = append(newDefinitions, newDef)
 		}
-	}
-	if len(newPartitions) == 0 {
-		job.State = model.JobStateCancelled
-		return ver, table.ErrUnknownPartition.GenWithStackByArgs(fmt.Sprintf("pid:%v", oldIDs), tblInfo.Name.O)
+		pi.OriginalDefinitions = pi.Definitions
+		pi.Definitions = newDefinitions
+
+		job.SchemaState = model.StateDeleteOnly
+		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, true)
+	case model.StateDeleteOnly:
+		// This state is not a real 'DeleteOnly' state, because tidb does not maintaining the state check in partitionDefinition.
+		// Insert this state to confirm all servers can not see the old partitions when reorg is running,
+		// so that no new data will be inserted into old partitions when reorganizing.
```

Comment:
Could add a comment here to indicates that the state of `DeleteOnly` is 'fake', and the insert into new partition can still be successful in this state.

Reply:
OK, I add it.